keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,956 @@
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+We intend to introduce a "compatibility version" option to Kubernetes control
+plane components to make upgrades safer by increasing the granularity of steps
+available to cluster administrators. The "compatibility version" is distinct from
+the binary version of a Kubernetes component, and can be used to emulate the
+behavior (APIs, features, ...) of a prior Kubernetes release version.
+
+## Motivation
+
+The notion of more granular steps in Kubernetes upgrades is attractive because
+it is more rigorous about how we step through a Kubernetes control-plane
+upgrade, introducing potentially corrupting data (i.e. data only present in N+1
+and not in N) only in later stages of the upgrade process.
+
+For example, upgrading from Kubernetes 1.30 to 1.31 while keeping the compatibility
+version at 1.30 would enable a cluster administrator to validate that the new
+Kubernetes binary version is working as desired before exposing any feature changes
+introduced in 1.31 to cluster users, and without writing and data to storage at
+newer API versions.
+
+This extra step increases the granularity of our upgrade sequence so that 
+(1) failures are more easily diagnosed (since we have more granular steps, we
+have slightly more data sheerly from which point of the upgrade lifecycle we
+achieved failure) and (2) failures are more easily auto-reverted by upgrade
+orchestration as we are taking smaller and more incremental steps forward,
+which means there is less to “undo” on a failure condition.
+
+In beta, we intend to introduce support for greater compatibility version skew
+(specifically, N-3) so that it would be possible to skip-level upgrade a 
+Kubernetes control-plane, by means of:
+
+- (starting point) binary-version 1.28 (compat-version 1.28)
+- upgrade binary-version to 1.31 (compat-version stays at 1.28 - this is our skip-level binary upgrade)
```

Comment:
i'm thinking about a situation this sequence may not work. (i may be incorrect, lmk)

Say featureX is GA in 1.27 and gets removed in 1.30 and featureY is GA in 1.28 and is removed in 1.31. This sequence would need users to update in the said step sequence from 1.28 to 1.30 only(because 1.30 to 1.31 compat version change may break something for the users. Is that a good way to think?

Reply:
I think everything you're saying about potential for breakage is true.

What we're trying to say in this line item (maybe unsuccessfully), is that you could upgrade in the following steps:

-  `binary version: 1.28, `compatibility-version: unset`
-  `binary version: 1.30`, `compatibility-version: 1.28`
-  `binary version: 1.30`, `compatibility-version: 1.29`
-  `binary version: 1.30`, `compatibility-version: 1.30`

Feature removal is a possibility at each step, so from that perspective, this is no better or worse than directly performing binary upgrades.  But it does mean that fewer binary versions need to be involved in the upgrade. This is attractive for anyone that wants to minimize the number of binaries involved in an upgrade.

Reply:
> This is attractive for anyone that wants to minimize the number of binary versions involved in an upgrade

This is true. Given you also said that from a failure perspective this is no better or worse than direct binary updates, could we mention a few failure scenarios that could benefit from this sequence? I ask because this could paint a better guidance on when to use this fashion.

Reply:
Sure.  I'll summarize here and then post into the KEP.

While failures due to feature/API removal are the same, the total number of possible upgrade steps changes.  The first step, where the binary version changes but the compatibility version does not change, offers an opportunity to verify the binary version before accomulating any client usage of features/API at the new version (since the compatibility version is unchanged).  This has a few benefits:

- Any upgrade system that successfully detects failures between upgrade steps can report which upgrade step failed. This makes it easier to diagnose the failures, because there are fewer possible causes of the failure.  (There's a huge difference between "A cluster upgrade failed" and "A cluster upgrade failed when only the apiserver's binary version changed").
- For upgrades where multiple failures occur, this increases the odds those failures are split across steps.  An upgrade system that is able to pause after a step where failures are detected allows cluster administrators to address failures caused by one step, without the disruption and "noise" cause the other failures from future steps. 

Both offer opportunities for lower MTTD and MTTR of failures.

Reply:
Thank you very much! These would be great to have in KEP Summary

Reply:
An additional point could be. A binary version upgrade is much more heavy. In a all or nothing world, the cluster operators have to really rollback the whole upgrade sequence. This fashion allows the operators to mitigate error by not rolling back the entire sequence, but mutate the compat flag only to debug. May be this point is implicit, but its a major hurdle at least in the system i operate in.